apps/gdal_contour.cpp

```diff
@@ -33,45 +33,193 @@
+    bool b3D = false;
+    bool bPolygonize = false;
+    bool bNoDataSet = false;
+    bool bIgnoreNoData = false;
+    std::string osNewLayerName = "contour";
+    std::string osFormat;
+    std::string osElevAttrib;
+    std::string osElevAttribMin;
+    std::string osElevAttribMax;
+    std::vector<double> adfFixedLevels;
+    CPLStringList aosOpenOptions;
+    CPLStringList aosCreationOptions;
+    bool bQuiet = false;
+    std::string aosDestFilename;
+    std::string aosSrcFilename;
+};
+
+/************************************************************************/
+/*                     GDALContourAppOptionsGetParser()                 */
+/************************************************************************/
 
+static std::unique_ptr<GDALArgumentParser>
+GDALContourAppOptionsGetParser(GDALContourOptions *psOptions)
 {
-    return CPLGetValueType(pszArg) != CPL_VALUE_STRING;
+    auto argParser = std::make_unique<GDALArgumentParser>(
+        "gdal_contour", /* bForBinary */ true);
+
+    argParser->add_description(_("Creates contour lines from a raster file."));
+    argParser->add_epilog(_(
+        "For more details, consult the full documentation for the gdal_contour "
+        "utility: http://gdal.org/gdal_contour.html"));
+
+    argParser->add_argument("-b")
+        .metavar("<name>")
+        .default_value(1)
+        .nargs(1)
+        .scan<'i', int>()
+        .store_into(psOptions->nBand)
+        .help(_("Select an input band band containing the DEM data."));
+
+    argParser->add_argument("-a")
+        .metavar("<name>")
+        .store_into(psOptions->osElevAttrib)
+        .help(_("Provides a name for the attribute in which to put the "
+                "elevation."));
+
+    argParser->add_argument("-amin")
+        .metavar("<name>")
+        .store_into(psOptions->osElevAttribMin)
+        .help(_("Provides a name for the attribute in which to put the minimum "
+                "elevation."));
+
+    argParser->add_argument("-amax")
+        .metavar("<name>")
+        .store_into(psOptions->osElevAttribMax)
+        .help(_("Provides a name for the attribute in which to put the maximum "
+                "elevation."));
+
+    argParser->add_argument("-3d")
+        .flag()
+        .store_into(psOptions->b3D)
+        .help(_("Force production of 3D vectors instead of 2D."));
+
+    argParser->add_argument("-inodata")
+        .flag()
+        .store_into(psOptions->bIgnoreNoData)
+        .help(_("Ignore any nodata value implied in the dataset - treat all "
+                "values as valid."));
+
+    argParser->add_argument("-snodata")
+        .metavar("<value>")
+        .scan<'g', double>()
+        .action(
+            [psOptions](const auto &d)
+            {
+                psOptions->bNoDataSet = true;
+                psOptions->dfNoData = CPLAtofM(d.c_str());
+            })
+        .help(_("Input pixel value to treat as \"nodata\"."));
+
+    argParser->add_argument("-f")
```

Comment:
you can use add_output_format_argument()

Reply:
done